From 6e68400fc04c4f056b08877ef1f0a61ffb47136d Mon Sep 17 00:00:00 2001
From: Min He <min.he@intel.com>
Date: Wed, 2 Mar 2016 22:26:44 +0800
Subject: [PATCH 373/403] Fix an logic issue in ppgtt_allocate_oos_page

In ppgtt_allocate_oos_page, when there's no free oos_page, we'll sync an
old oos page and detach it. But that oos page may not belong to current
vgt, so we need to find corresponding vgt and sync to it.

This patch can fix bug 1005.

Signed-off-by: Min He <min.he@intel.com>
---
 drivers/gpu/drm/i915/vgt/gtt.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/vgt/gtt.c b/drivers/gpu/drm/i915/vgt/gtt.c
index 6e3d431..3dc6086 100644
--- a/drivers/gpu/drm/i915/vgt/gtt.c
+++ b/drivers/gpu/drm/i915/vgt/gtt.c
@@ -1179,13 +1179,15 @@ static bool ppgtt_allocate_oos_page(struct vgt_device *vgt, guest_page_t *gpt)
 	struct pgt_device *pdev = vgt->pdev;
 	struct vgt_gtt_info *gtt = &pdev->gtt;
 	oos_page_t *oos_page = gpt->oos_page;
+	ppgtt_spt_t *spt = NULL;
 
 	/* oos_page should be NULL at this point */
 	ASSERT(!oos_page);
 
 	if (list_empty(&gtt->oos_page_free_list_head)) {
 		oos_page = container_of(gtt->oos_page_use_list_head.next, oos_page_t, list);
-		if (!ppgtt_set_guest_page_sync(vgt, oos_page->guest_page)
+		spt = guest_page_to_ppgtt_spt(oos_page->guest_page);
+		if (!ppgtt_set_guest_page_sync(spt->vgt, oos_page->guest_page)
 			|| !vgt_detach_oos_page(vgt, oos_page))
 			return false;
 		ASSERT(!list_empty(&gtt->oos_page_free_list_head));
-- 
1.7.10.4

